racket/collects/racket/match/parse.rkt

```diff
@@ -195,6 +195,39 @@
        (raise-syntax-error #f "key does not have a value" stx)]
       [else (loop (rest (rest xs)) (cons (list (first xs) (second xs)) acc))])))
 
+(define (impossible-and? list-of-patterns)
+  (define ps list-of-patterns)
+  ; Patterns that contain two different literals such
+  ; as (and 10 11) can never match a value.
+  ; Such a pattern can be replaced with the pattern (or).
+
+  (define (literal? x)
+    (or (eq? x #t)
+        (eq? x #f)
+        (string? x)
+        (bytes? x)
+        (number? x)
+        (char? x)
+        (keyword? x)
+        (regexp? x)
+        (pregexp? x)))
+
+  (define (stxliteral? x)
+    (literal? (if (syntax? x) (syntax-e x) x)))
+
+  (define (same? x y)
+    (equal? (if (syntax? x) (syntax-e x) x)
+            (if (syntax? y) (syntax-e y) y)))
+
+  (define all-literals (map syntax-e (filter stxliteral? ps)))
```

Comment:
This doesn't look right to me. Does it preserve side-effect on the following program?

```
#lang racket

(define (f x)
  (match x
    [(and (? (begin (println 'hello) values))
          1
          2)
     3]
    [_ 4]))

(f 10)
```